qmetaobject/src/core.rs

```diff
@@ -0,0 +1,129 @@
+/* Copyright (C) 2018 Olivier Goffart <ogoffart@woboq.com>
+   Copyright (C) 2021 ivan tkachenko a.k.a. ratijas <me@ratijas.tk>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
+associated documentation files (the "Software"), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial
+portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+//! Wrappers around QtCore module things.
+
+use cpp::cpp;
+
+use crate::*;
+
+cpp! {{
+    #include <QtCore/QCoreApplication>
+}}
+
+/// Wrapper around [`QCoreApplication`][qt] class.
+///
+/// # Wrapper-specific
+///
+/// Currently it is uncreatible, non-obtainable, and exists purely as a
+/// namespace for public static members (associated functions in Rust).
+///
+/// [qt]: https://doc.qt.io/qt-5/qcoreapplication.html
+pub struct QCoreApplication {
+    // Private field makes this dummy struct uncreatable by user. Since
+    // there's no way to obtain an instance of it, we won't have to worry
+    // about layout compatibility and stuff.
+    _private: (),
+}
+
+impl QCoreApplication {
+    /// Wrapper around [`QString applicationName()`][qt] static method.
+    ///
+    /// [qt]: https://doc.qt.io/qt-5/qcoreapplication.html#applicationName-prop
+    pub fn application_name() -> String {
+        cpp!(unsafe [] -> QString as "QString" {
+            return QCoreApplication::applicationName();
+        }).to_string()
+    }
+
+    /// Wrapper around [`void setApplicationName(const QString &application)`][qt] static method.
+    ///
+    /// [qt]: https://doc.qt.io/qt-5/qcoreapplication.html#applicationName-prop
+    pub fn set_application_name(application: &str) {
```

Comment:
Why not take a QString since we take ownership as a QString?

Reply:
AFAIK String to QString conversion is not in-place, and re-allocated memory. So, why drop a String if we only need a str slice? This follows minimal requirements rule.

Here are the conversion impls, btw
```rust
impl<'a> From<&'a str> for QString {
    /// Copy the data from a `&str`.
    fn from(s: &'a str) -> QString {
        let len = s.len();
        let ptr = s.as_ptr();
        cpp!(unsafe [len as "size_t", ptr as "char*"] -> QString as "QString" {
            return QString::fromUtf8(ptr, len);
        })
    }
}
impl From<String> for QString {
    fn from(s: String) -> QString {
        QString::from(&*s)
    }
}
```

Reply:
Hilariously, documentation of [QString::fromUtf8](https://doc.qt.io/qt-5/qstring.html#fromUtf8) says nothing about ownership. It just "returns a QString initialized with the first size bytes of the UTF-8 string str" — whatever that "initialization" might imply. And the type of `const char *str` requiring it being const doesn't make the whole situation better. Here we go reading source code once again…

Reply:
I think, the broad question here is: do we want to make wrappers more "Rust idiomatic" by exposing QString as String/&str etc., or do we want preserve the implicit sharing provided by Qt data types (not only strings)?

Reply:
Right, QString is owning the string.    So it allocates.  Just like `String` in rust.

> So, why drop a String if we only need a str slice?

We don't need a str slice,  we need a QString to give it to the application.

Reply:
Shout-outs to #81, by the way. I think it was mentioned somewhere there, but was never resolved. If not, then it should be added.

So, give me a second to refactor it into QStrings.